File: 3665.cpp
#include <bits/stdc++.h>
using namespace std;

constexpr size_t MAX_N = 501;

bool arr[MAX_N][MAX_N];
int income[MAX_N];
int t[MAX_N];
int tc, n, m, a, b;

void solve()
{
    stack<int> s;

    for (int i = 1; i <= n; i++)
        if (income[i] == 0)
            s.push(i);

    vector<int> ans;
    bool uncertain = false;

    while (!s.empty())
    {
        if (s.size() > 1)
            uncertain = true;

        int now = s.top();

        ans.push_back(now);
        s.pop();

        for (int i = 1; i <= n; i++)
        {
            if (!arr[now][i])
                continue;

            if (--income[i] == 0)
                s.push(i);
        }
    }

    if (ans.size() < n)
        cout << "IMPOSSIBLE\n";
    else if (uncertain)
        cout << "?\n";
    else
    {
        for (const int i: ans)
            cout << i << ' ';
        cout << '\n';
    }
}

int main()
{
    cin.tie(nullptr)->sync_with_stdio(false);

    for (cin >> tc; tc--;)
    {
        memset(arr, false, sizeof arr);
        memset(income, 0, sizeof income);

        cin >> n;
        for (int i = 1; i <= n; i++)
        {
            cin >> t[i];

            for (int j = 1; j < i; j++)
                arr[t[j]][t[i]] = true;
            income[t[i]] = i - 1;
        }

        for (cin >> m; m--;)
        {
            cin >> a >> b;
            if (arr[b][a])
                swap(a, b);

            arr[a][b] = false;
            arr[b][a] = true;
            income[b]--;
            income[a]++;
        }

        solve();
    }

    return 0;
}

File: 3665.py
import sys
input = sys.stdin.readline

def solve():
    adj = [[False] * (n + 1) for _ in range(n + 1)]
    inc = [0] * (n + 1)

    t_len = len(t)
    for i in range(t_len):
        for j in range(i + 1, t_len):
            adj[t[j]][t[i]] = True
            inc[t[i]] += 1

    for a, b in ab:
        if not adj[a][b]:
            a, b = b, a

        adj[a][b] = False
        adj[b][a] = True
        inc[b] -= 1
        inc[a] += 1

    v = []

    for i in range(1, n + 1):
        if inc[i] == 0:
            v.append(i)

    ans = []
    uncertain = False

    while v:
        if len(v) > 1:
            uncertain = True

        now = v.pop()
        ans.append(now)

        for i in range(1, n + 1):
            if not adj[now][i]:
                continue

            inc[i] -= 1
            if inc[i] == 0:
                v.append(i)

    if len(ans) < n:
        return ['IMPOSSIBLE']

    if uncertain:
        return ['?']

    return ans[::-1]

for _ in range(int(input())):
    n = int(input())
    t = [*map(int, input().split())]
    ab = [[*map(int, input().split())] for _ in range(int(input()))]

    print(*solve())
